packages/flutter/lib/src/material/text_field.dart

```diff
@@ -1048,10 +1073,18 @@ class _TextFieldState extends State<TextField> with RestorationMixin implements
     }
   }
 
+  void _updateMaterialErrorState() {
+    if (_hasError != _hadError) {
+      _statesController.update(MaterialState.error, _hasError);
+      _hadError = _hasError;
+    }
+  }
+
   @override
   void didChangeDependencies() {
     super.didChangeDependencies();
     _effectiveFocusNode.canRequestFocus = _canRequestFocus;
+    _updateMaterialErrorState();
```

Comment:
`_hasError` doesn't seem to be depending on any inherited widgets?

Reply:
The `RestorableTextEditingController` doesn't seem to finish registering the controller by the time `_hasError` is called which also calls `_hasIntrinsicError` so when `effectiveController.value` is accessed there is an assert triggered regarding the controller not being registered for state restoration yet.

Reply:
You mean it's preventing the material state controller from being fully set-up in `initState`? Can you make the initialization lazy then?

Reply:
Making the initialization lazy didn't seem to help, I tried changing the MaterialStatesController to `late`. Instead I changed `_hasIntrinsicError` to take into account if the RestorableTextEditingController has been registered.